### ISSUE TITLE
bufferTime operator is impure

### ISSUE THREAD

**Issue by joshribakoff** [ReactiveX/rxjs#5528]
## Bug Report

**Current Behavior**

```
import { interval, Subject } from 'rxjs'; 
import { bufferTime, takeWhile, tap } from 'rxjs/operators';


const subject = new Subject<number>()

interval(100).pipe(tap(n => subject.next(n))).subscribe();

const log = batch => {
  // setTimeout(()=>{
    subject.next(99)
  // })
  console.log(batch)
}

const source = subject.pipe(
  takeWhile(x => x < 100),
  bufferTime(1000),
  tap(batch => log(batch))
);

source.subscribe();
```

`99` shows up at end of each array that bufferTime emits, it should not be possible for an operator that comes later to modify a value a previous operator emits. It should show up at the start of the 2nd buffer onwards.

Discovered while unit testing code that pushes onto a buffer in a jest mock callback that was being used in a side effect, after a bufferTime().

**Environment**
- Runtime: both node / chrome
- RxJS version: 6.5.4

**Possible Solution**
don't mutate

**Comment by cartant** [comment]
Yeah, this is a bug. `buffer` doesn't have this problem, as it assigns a new array to the subscriber's property before nexting:

https://github.com/ReactiveX/rxjs/blob/82bcc2179b8aefc08342a419ccc549080c358a64/src/internal/operators/buffer.ts#L84-L86

`bufferTime` does not. This code needs to be rearranged:

https://github.com/ReactiveX/rxjs/blob/82bcc2179b8aefc08342a419ccc549080c358a64/src/internal/operators/bufferTime.ts#L212-L218

**Comment by joshribakoff** [comment]
Could I volunteer to tackle the fix & create the PR?

**Comment by cartant** [comment]
> Could I volunteer to tackle the fix & create the PR?

Sure. If you could create a PR that contains two commits - a failing test that reproduces the problem; and the fix - that would be great. Thanks.

**Comment by joshribakoff** [comment]
This is done. It was a bit more complex than just swapping the order of statements in the code snippet you posted above, unfortunately. I can see some refactoring I'd like to discuss separately after I get my feet more wet, will post more follow up Qs on the PR.
